Auto-installed Node.js in .vaadin should not take precedence over global by default

Flow downloads and installs Node.js into `~/.vaadin` if it can't find `npm` command globally (i.e., on `PATH`). After such auto-installation, Flow will subsequently prefer `npm` and `node` commands in  `~/.vaadin`, even if a global installation has been added later. This version of node is used both when Flow runs `npm install` (when `vaadin.pnpm.enable=false`) and when it launches `webpack-dev-server`. 

This is contradictory to the auto-install logic (which prefers global) and should not be the default behaviour.  Only if the `require.home.node` parameter is set to true (false by default) should the `~/.vaadin/node/node` be forced.

Flow version: 6.0-SNAPSHOT (2021-01-22).

Also seem to happen in 14.4.6. I keep deleting ~/.vaadin/node and it works for a period of time. Then all of a sudden builds will fail with this message:

`
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:14.4.6:prepare-frontend (default) on project bookingcentralx: 
[ERROR] 
[ERROR] ======================================================================================================
[ERROR] Your installed 'node' version (8.1.2) is too old. Supported versions are 10.0+
[ERROR] Please install a new one either:
[ERROR]   - by following the https://nodejs.org/en/download/ guide to install it globally
[ERROR]   - or by running the frontend-maven-plugin goal to install it in this project:
[ERROR]   $ mvn com.github.eirslett:frontend-maven-plugin:1.7.6:install-node-and-npm -DnodeVersion="v12.14.0" 
[ERROR] 
[ERROR] You can disable the version check using -Dvaadin.ignoreVersionChecks=true
[ERROR] ======================================================================================================
[ERROR] 
[ERROR] -> [Help 1]
`

--- a/flow-server/src/main/java/com/vaadin/flow/server/DevModeHandler.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/DevModeHandler.java
@@ -659,12 +659,12 @@ public final class DevModeHandler implements RequestHandler {
         ProcessBuilder processBuilder = new ProcessBuilder()
                 .directory(npmFolder);
 
-        FrontendTools tools = new FrontendTools(npmFolder.getAbsolutePath(),
-                () -> FrontendUtils.getVaadinHomeDirectory().getAbsolutePath());
-        tools.validateNodeAndNpmVersion();
-
         boolean useHomeNodeExec = config.getBooleanProperty(
                 InitParameters.REQUIRE_HOME_NODE_EXECUTABLE, false);
+        FrontendTools tools = new FrontendTools(npmFolder.getAbsolutePath(),
+                () -> FrontendUtils.getVaadinHomeDirectory().getAbsolutePath(),
+                useHomeNodeExec);
+        tools.validateNodeAndNpmVersion();
 
         String nodeExec = null;
         if (useHomeNodeExec) {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
@@ -25,6 +25,7 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.Properties;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
@@ -622,11 +623,11 @@ public class FrontendTools {
         }
         if (returnCommand.isEmpty()) {
             // Otherwise look for regular `npm`/`npx` global search path
-            String command = frontendToolsLocator
+            Optional<String> command = frontendToolsLocator
                     .tryLocateTool(cliTool.getCommand())
-                    .map(File::getAbsolutePath).orElse(null);
-            if (command != null) {
-                returnCommand = Collections.singletonList(command);
+                    .map(File::getAbsolutePath);
+            if (command.isPresent()) {
+                returnCommand = Collections.singletonList(command.get());
             }
         }
         if (!alternativeDirChecked && returnCommand.isEmpty()) {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
@@ -622,9 +622,12 @@ public class FrontendTools {
         }
         if (returnCommand.isEmpty()) {
             // Otherwise look for regular `npm`/`npx` global search path
-            returnCommand = Collections.singletonList(
-                    frontendToolsLocator.tryLocateTool(cliTool.getCommand())
-                            .orElse(null).getAbsolutePath());
+            String command = frontendToolsLocator
+                    .tryLocateTool(cliTool.getCommand())
+                    .map(File::getAbsolutePath).orElse(null);
+            if (command != null) {
+                returnCommand = Collections.singletonList(command);
+            }
         }
         if (!alternativeDirChecked && returnCommand.isEmpty()) {
             // Use alternative if global is not found and alternative location

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
@@ -229,7 +229,7 @@ public class FrontendTools {
      *            provided from an intranet mirror. Use
      *            {@link NodeInstaller#DEFAULT_NODEJS_DOWNLOAD_ROOT} by default.
      * @param forceAlternativeNode
-     *            force alternative node executable using
+     *            force usage of node executable from alternative directory
      */
     public FrontendTools(String baseDir, Supplier<String> alternativeDirGetter,
             String nodeVersion, URI nodeDownloadRoot,

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
@@ -154,6 +154,33 @@ public class FrontendTools {
 
     private final boolean ignoreVersionChecks;
 
+    private final boolean forceAlternativeNode;
+
+    /**
+     * Creates an instance of the class using the {@code baseDir} as a base
+     * directory to locate the tools and the directory returned by the
+     * {@code alternativeDirGetter} as a directory to install tools if they are
+     * not found and use it as an alternative tools location.
+     * <p>
+     * If {@code alternativeDir} is {@code null} tools won't be installed.
+     *
+     *
+     * @param baseDir
+     *            the base directory to locate the tools, not {@code null}
+     * @param alternativeDirGetter
+     *            the getter for a directory where tools will be installed if
+     *            they are not found globally or in the {@code baseDir}, may be
+     *            {@code null}
+     * @param forceAlternativeNode
+     *            force alternative node executable using
+     */
+    public FrontendTools(String baseDir, Supplier<String> alternativeDirGetter,
+            boolean forceAlternativeNode) {
+        this(baseDir, alternativeDirGetter, DEFAULT_NODE_VERSION,
+                URI.create(NodeInstaller.DEFAULT_NODEJS_DOWNLOAD_ROOT),
+                forceAlternativeNode);
+    }
+
     /**
      * Creates an instance of the class using the {@code baseDir} as a base
      * directory to locate the tools and the directory returned by the
@@ -173,7 +200,43 @@ public class FrontendTools {
     public FrontendTools(String baseDir,
             Supplier<String> alternativeDirGetter) {
         this(baseDir, alternativeDirGetter, DEFAULT_NODE_VERSION,
-                URI.create(NodeInstaller.DEFAULT_NODEJS_DOWNLOAD_ROOT));
+                URI.create(NodeInstaller.DEFAULT_NODEJS_DOWNLOAD_ROOT), false);
+    }
+
+    /**
+     * Creates an instance of the class using the {@code baseDir} as a base
+     * directory to locate the tools and the directory returned by the
+     * {@code alternativeDirGetter} as a directory to install tools if they are
+     * not found and use it as an alternative tools location.
+     * <p>
+     * If {@code alternativeDir} is {@code null} tools won't be installed.
+     *
+     *
+     * @param baseDir
+     *            the base directory to locate the tools, not {@code null}
+     * @param alternativeDirGetter
+     *            the getter for a directory where tools will be installed if
+     *            they are not found globally or in the {@code baseDir}, may be
+     *            {@code null}
+     * @param nodeVersion
+     *            The node.js version to be used when node.js is installed
+     *            automatically by Vaadin, for example <code>"v14.15.4"</code>.
+     *            Use {@value #DEFAULT_NODE_VERSION} by default.
+     * @param nodeDownloadRoot
+     *            Download node.js from this URL. Handy in heavily firewalled
+     *            corporate environments where the node.js download can be
+     *            provided from an intranet mirror. Use
+     *            {@link NodeInstaller#DEFAULT_NODEJS_DOWNLOAD_ROOT} by default.
+     * @param forceAlternativeNode
+     *            force alternative node executable using
+     */
+    public FrontendTools(String baseDir, Supplier<String> alternativeDirGetter,
+            String nodeVersion, URI nodeDownloadRoot,
+            boolean forceAlternativeNode) {
+        this(baseDir, alternativeDirGetter, nodeVersion, nodeDownloadRoot,
+                "true".equalsIgnoreCase(System.getProperty(
+                        FrontendUtils.PARAM_IGNORE_VERSION_CHECKS)),
+                forceAlternativeNode);
     }
 
     /**
@@ -204,18 +267,18 @@ public class FrontendTools {
     public FrontendTools(String baseDir, Supplier<String> alternativeDirGetter,
             String nodeVersion, URI nodeDownloadRoot) {
         this(baseDir, alternativeDirGetter, nodeVersion, nodeDownloadRoot,
-                "true".equalsIgnoreCase(System.getProperty(
-                        FrontendUtils.PARAM_IGNORE_VERSION_CHECKS)));
+                false);
     }
 
     FrontendTools(String baseDir, Supplier<String> alternativeDirGetter,
             String nodeVersion, URI nodeDownloadRoot,
-            boolean ignoreVersionChecks) {
+            boolean ignoreVersionChecks, boolean forceAlternativeNode) {
         this.baseDir = Objects.requireNonNull(baseDir);
         this.alternativeDirGetter = alternativeDirGetter;
         this.nodeVersion = Objects.requireNonNull(nodeVersion);
         this.nodeDownloadRoot = Objects.requireNonNull(nodeDownloadRoot);
         this.ignoreVersionChecks = ignoreVersionChecks;
+        this.forceAlternativeNode = forceAlternativeNode;
     }
 
     /**
@@ -225,8 +288,23 @@ public class FrontendTools {
      */
     public String getNodeExecutable() {
         Pair<String, String> nodeCommands = getNodeCommands();
-        return getExecutable(nodeCommands.getFirst(), nodeCommands.getSecond(),
-                alternativeDirGetter != null).getAbsolutePath();
+        File file = getExecutable(baseDir, nodeCommands.getSecond());
+        if (file == null) {
+            file = frontendToolsLocator.tryLocateTool(nodeCommands.getFirst())
+                    .orElse(null);
+        }
+        if (file == null) {
+            file = getExecutable(getAlternativeDir(), nodeCommands.getSecond());
+        }
+        if (file == null && alternativeDirGetter != null) {
+            getLogger().info("Couldn't find {}. Installing Node and NPM to {}.",
+                    nodeCommands.getFirst(), getAlternativeDir());
+            file = new File(installNode(nodeVersion, nodeDownloadRoot));
+        }
+        if (file == null) {
+            throw new IllegalStateException(String.format(NODE_NOT_FOUND));
+        }
+        return file.getAbsolutePath();
     }
 
     /**
@@ -299,7 +377,7 @@ public class FrontendTools {
         }
         try {
             List<String> nodeVersionCommand = new ArrayList<>();
-            nodeVersionCommand.add(getNodeExecutable());
+            nodeVersionCommand.add(doGetNodeExecutable());
             nodeVersionCommand.add("--version"); // NOSONAR
             FrontendVersion foundNodeVersion = FrontendUtils.getVersion("node",
                     nodeVersionCommand);
@@ -385,28 +463,12 @@ public class FrontendTools {
         }
     }
 
-    private File getExecutable(String cmd, String defaultLocation,
-            boolean installNode) {
-        File file = null;
-        if (defaultLocation == null) {
-            file = frontendToolsLocator.tryLocateTool(cmd).orElse(null);
-        } else {
-            file = Arrays.asList(() -> baseDir, alternativeDirGetter).stream()
-                    .map(Supplier::get)
-                    .map(dir -> new File(dir, defaultLocation))
-                    .filter(frontendToolsLocator::verifyTool).findFirst()
-                    .orElseGet(() -> frontendToolsLocator.tryLocateTool(cmd)
-                            .orElse(null));
+    private File getExecutable(String dir, String location) {
+        File file = new File(dir, location);
+        if (frontendToolsLocator.verifyTool(file)) {
+            return file;
         }
-        if (file == null && installNode) {
-            getLogger().info("Couldn't find {}. Installing Node and NPM to {}.",
-                    cmd, getAlternativeDir());
-            return new File(installNode(nodeVersion, nodeDownloadRoot));
-        }
-        if (file == null) {
-            throw new IllegalStateException(String.format(NODE_NOT_FOUND));
-        }
-        return file;
+        return null;
     }
 
     private Pair<String, String> getNodeCommands() {
@@ -550,17 +612,27 @@ public class FrontendTools {
         // First look for *-cli.js script in project/node_modules
         List<String> returnCommand = getNpmScriptCommand(baseDir,
                 cliTool.getScript());
-        if (returnCommand.isEmpty()) {
+        boolean alternativeDirChecked = false;
+        if (returnCommand.isEmpty() && forceAlternativeNode) {
             // First look for *-cli.js script in ~/.vaadin/node/node_modules
+            // only if alternative node takes precedence over all other location
             returnCommand = getNpmScriptCommand(getAlternativeDir(),
                     cliTool.getScript());
+            alternativeDirChecked = true;
         }
         if (returnCommand.isEmpty()) {
             // Otherwise look for regular `npm`/`npx` global search path
             returnCommand = Collections.singletonList(
-                    getExecutable(cliTool.getCommand(), null, true)
-                            .getAbsolutePath());
+                    frontendToolsLocator.tryLocateTool(cliTool.getCommand())
+                            .orElse(null).getAbsolutePath());
         }
+        if (!alternativeDirChecked && returnCommand.isEmpty()) {
+            // Use alternative if global is not found and alternative location
+            // is not yet checked
+            returnCommand = getNpmScriptCommand(getAlternativeDir(),
+                    cliTool.getScript());
+        }
+
         if (flags.length > 0) {
             returnCommand = new ArrayList<>(returnCommand);
             Collections.addAll(returnCommand, flags);
@@ -576,7 +648,7 @@ public class FrontendTools {
         List<String> returnCommand = new ArrayList<>();
         if (file.canRead()) {
             // We return a two element list with node binary and npm-cli script
-            returnCommand.add(getNodeExecutable());
+            returnCommand.add(doGetNodeExecutable());
             returnCommand.add(file.getAbsolutePath());
         }
         return returnCommand;
@@ -638,5 +710,13 @@ public class FrontendTools {
 
     private String getAlternativeDir() {
         return alternativeDirGetter.get();
+    }
+
+    private String doGetNodeExecutable() {
+        if (forceAlternativeNode) {
+            return forceAlternativeNodeExecutable();
+        } else {
+            return getNodeExecutable();
+        }
     }
 }

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
@@ -173,7 +173,7 @@ public class FrontendTools {
      *            they are not found globally or in the {@code baseDir}, may be
      *            {@code null}
      * @param forceAlternativeNode
-     *            force alternative node executable using
+     *            force usage of node executable from alternative directory
      */
     public FrontendTools(String baseDir, Supplier<String> alternativeDirGetter,
             boolean forceAlternativeNode) {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
@@ -375,7 +375,7 @@ public class TaskRunNpmInstall implements FallibleCommand {
 
         FrontendTools tools = new FrontendTools(baseDir,
                 () -> FrontendUtils.getVaadinHomeDirectory().getAbsolutePath(),
-                nodeVersion, nodeDownloadRoot);
+                nodeVersion, nodeDownloadRoot, requireHomeNodeExec);
         try {
             if (requireHomeNodeExec) {
                 tools.forceAlternativeNodeExecutable();

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java
@@ -438,6 +438,7 @@ public class FrontendToolsTest {
         Assume.assumeFalse(
                 "Skipping test on windows until a fake node.exe that isn't caught by Window defender can be created.",
                 FrontendUtils.isWindows());
+        tools = new FrontendTools(baseDir, () -> vaadinHomeDir, true);
         assertNodeCommand(() -> vaadinHomeDir);
     }
 
@@ -456,6 +457,7 @@ public class FrontendToolsTest {
         Assume.assumeFalse(
                 "Skipping test on windows until a fake node.exe that isn't caught by Window defender can be created.",
                 FrontendUtils.isWindows());
+        tools = new FrontendTools(baseDir, () -> vaadinHomeDir, true);
         assertNpmCommand(() -> vaadinHomeDir);
     }
 

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java
@@ -190,12 +190,17 @@ public class FrontendToolsTest {
     }
 
     @Test
-    public void should_useSystemNode() {
+    public void homeNodeIsNotForced_useGlobalNode() throws IOException {
+        createStubNode(true, true, vaadinHomeDir);
+
         assertThat(tools.getNodeExecutable(), containsString("node"));
         assertThat(tools.getNodeExecutable(),
                 not(containsString(DEFAULT_NODE)));
         assertThat(tools.getNodeExecutable(),
                 not(containsString(NPM_CLI_STRING)));
+        assertThat(tools.getNodeExecutable(),
+                not(containsString(vaadinHomeDir)));
+        assertThat(tools.getNodeExecutable(), not(containsString(baseDir)));
 
         assertEquals(4, tools.getNpmExecutable().size());
         assertThat(tools.getNpmExecutable().get(0), containsString("npm"));
@@ -434,15 +439,6 @@ public class FrontendToolsTest {
     }
 
     @Test
-    public void should_useHomeFirst() throws Exception {
-        Assume.assumeFalse(
-                "Skipping test on windows until a fake node.exe that isn't caught by Window defender can be created.",
-                FrontendUtils.isWindows());
-        tools = new FrontendTools(baseDir, () -> vaadinHomeDir, true);
-        assertNodeCommand(() -> vaadinHomeDir);
-    }
-
-    @Test
     public void should_useProjectNpmFirst() throws Exception {
         Assume.assumeFalse(
                 "Skipping test on windows until a fake node.exe that isn't caught by Window defender can be created.",
@@ -453,11 +449,13 @@ public class FrontendToolsTest {
     }
 
     @Test
-    public void should_useHomeNpmFirst() throws Exception {
+    public void forceHomeNode_useHomeNpmFirst() throws Exception {
         Assume.assumeFalse(
                 "Skipping test on windows until a fake node.exe that isn't caught by Window defender can be created.",
                 FrontendUtils.isWindows());
         tools = new FrontendTools(baseDir, () -> vaadinHomeDir, true);
+
+        createStubNode(true, true, vaadinHomeDir);
         assertNpmCommand(() -> vaadinHomeDir);
     }
 
@@ -487,7 +485,7 @@ public class FrontendToolsTest {
             throws Exception {
         tools = new FrontendTools(baseDir, () -> vaadinHomeDir,
                 FrontendTools.DEFAULT_NODE_VERSION, new File(baseDir).toURI(),
-                true);
+                true, false);
         Assume.assumeFalse(tools.getNodeExecutable().isEmpty());
         createStubNode(false, true, baseDir);
         createFakePnpm("4.5.0");

